nav2_docking/opennav_docking/src/controller.cpp

```diff
@@ -55,16 +80,98 @@ Controller::Controller(const rclcpp_lifecycle::LifecycleNode::SharedPtr & node)
   // Add callback for dynamic parameters
   dyn_params_handler_ = node->add_on_set_parameters_callback(
     std::bind(&Controller::dynamicParametersCallback, this, std::placeholders::_1));
+
+  node->get_parameter("controller.use_collision_detection", use_collision_detection_);
+  node->get_parameter("controller.projection_time", projection_time_);
+  node->get_parameter("controller.simulation_time_step", simulation_time_step_);
+
+  if (use_collision_detection_) {
+    double transform_tolerance;
+    std::string costmap_topic, footprint_topic;
+    node->get_parameter("controller.costmap_topic", costmap_topic);
+    node->get_parameter("controller.footprint_topic", footprint_topic);
+    node->get_parameter("controller.transform_tolerance", transform_tolerance);
+    node->get_parameter("controller.collision_tolerance", collision_tolerance_);
+    configureCollisionChecker(node, costmap_topic, footprint_topic, transform_tolerance);
+  }
+
+  trajectory_pub_ =
+    node->create_publisher<nav_msgs::msg::Path>(node->get_name() + std::string("/trajectory"), 1);
+}
+
+Controller::~Controller()
+{
+  control_law_.reset();
+  trajectory_pub_.reset();
+  costmap_sub_.reset();
+  footprint_sub_.reset();
+  collision_checker_.reset();
 }
 
 bool Controller::computeVelocityCommand(
   const geometry_msgs::msg::Pose & pose, geometry_msgs::msg::Twist & cmd, bool backward)
 {
   std::lock_guard<std::mutex> lock(dynamic_params_lock_);
   cmd = control_law_->calculateRegularVelocity(pose, backward);
+  return isTrajectoryCollisionFree(pose, backward);
+}
+
+bool Controller::isTrajectoryCollisionFree(
+  const geometry_msgs::msg::Pose & target_pose, bool backward)
+{
+  // Visualization of the trajectory
+  nav_msgs::msg::Path trajectory;
+  trajectory.header.frame_id = base_frame_;
+
+  // First pose
+  geometry_msgs::msg::PoseStamped next_pose;
+  next_pose.header.frame_id = base_frame_;
+  next_pose.pose.orientation.w = 1.0;
+  trajectory.poses.push_back(next_pose);
+
+  // Generate path
+  for (double t = 0; t < projection_time_; t += simulation_time_step_) {
+    // Apply velocities to calculate next pose
+    next_pose.pose = control_law_->calculateNextPose(
+      simulation_time_step_, target_pose, next_pose.pose, backward);
+
+    // Add the pose to the trajectory for visualization
+    trajectory.poses.push_back(next_pose);
+
+    // Transform pose from base_frame into fixed_frame
+    geometry_msgs::msg::PoseStamped local_pose = next_pose;
+    local_pose.header.stamp = rclcpp::Time(0);
+    tf2_buffer_->transform(local_pose, local_pose, fixed_frame_);
+
+    // Check for collisions at the projected pose
+    auto projected_pose = nav_2d_utils::poseToPose2D(local_pose.pose);
+    double distance = nav2_util::geometry_utils::euclidean_distance(target_pose, next_pose.pose);
+    if (use_collision_detection_ &&
+      !collision_checker_->isCollisionFree(projected_pose) && distance > collision_tolerance_)
```

Comment:
I don't understand the role of `collision_tolerance_` - can you explain this `0.3` for both the docking and undocking case?

Reply:
Does this properly handle the dock (end of trajectory stop checking) vs undock (start of trajectory don't check) cases? For undocking, we want to check the full trajectory to the prestaging pose, except the start segment

Reply:
I think this is what I remarked upon in my review main comment -- I think this is implementing what I think is missing, just needs some tweaks :-) 

I'd rename it too to `dock_collision_distance` --> to be clear this is the distance to dock collision to stop checking.

Reply:
I think this is the last major concern!

Reply:
Currently, it skip the last part of the trajectory for docking and undocking. Although for undocking it should check the whole trajectory up to the staging pose.

I made a quick video yesterday morning. It detects the obstacle, back to the staging pose. When the obstacle is removed, dock.
[docking_obstacle_bueno.webm](https://github.com/user-attachments/assets/95a44185-45ab-4a72-8aed-0dd3c537ccb7)

Reply:
> Currently, it skips the last part of the trajectory for docking and undocking. Although for undocking it should check the whole trajectory from the dock up to the staging pose.

I think there are different ways it should be handled:

Docking: From current pose --> Dock minus some tolerance near the dock (bc we're making contact)

Undocking: From current pose minus some tolerance (bc we're currently in contact) --> Undocking pose

I think the handling for docking vs undocking is inverted. Docking we ignore the *ending* bit for collision checking. Undocking we ignore the *starting* bit for collision checking, since we start presumably in collision with the dock.

Though -- I suppose we could parameterize whether we ignore anything. Presumably some infrastructure docking might not make contact, so we can collision check the full thing